File: Models/Stics_soil_temperature/Layerstemp.f90
MODULE layers_temp_mod
   IMPLICIT NONE
CONTAINS
   !%%CyML Model Begin%%
   SUBROUTINE model_layers_temp(temp_profile, &
                               layer_thick, &
                               layer_temp)

      IMPLICIT NONE

      REAL, INTENT(IN)    ::  temp_profile(:)
      INTEGER, INTENT(IN) ::  layer_thick(:)
      REAL, allocatable, INTENT(OUT)   ::  layer_temp(:)

      !%%CyML Description Begin%%
      !- Name: layers_temp -Version: 1.0, -Time step: 1
      !- Description:
      !            * Title: layers mean temperature model
      !            * Author: STICS
      !            * Reference: doi:http://dx.doi.org/10.1016/j.agrformet.2014.05.002
      !            * Institution: INRAE
      !            * Abstract: Calculates soil layers mean temperature
      !- inputs:
      !            * name: temp_profile
      !                          ** description : soil temperature profile
      !                          ** inputtype : variable
      !                          ** variablecategory : state
      !                          ** datatype : DOUBLEARRAY
      !                          ** default : 0.0
      !                          ** min : -50.0
      !                          ** max : 50.0
      !                          ** unit : degC
      !                          ** uri :
      !                          ** len : 
      !            * name: layer_thick
      !                          ** description : layers thickness
      !                          ** inputtype : parameter
      !                          ** parametercategory : constant
      !                          ** datatype : INTARRAY
      !                          ** default :
      !                          ** min :
      !                          ** max :
      !                          ** unit : cm
      !                          ** uri :
      !                          ** len : 
      !- outputs:
      !            * name: layer_temp
      !                          ** description : soil layers temperature
      !                          ** variablecategory : state
      !                          ** datatype : DOUBLEARRAY
      !                          ** min : -50.0
      !                          ** max : 50.0
      !                          ** unit : degC
      !                          ** uri :
      !                          ** len : 
      !%%CyML Description End%%

      INTEGER :: z
      INTEGER :: layers_nb
      INTEGER, allocatable :: up_depth(:)
      INTEGER, allocatable :: layer_depth(:)
      INTEGER :: depth_value

      !%%CyML Compute Begin%%
      call get_layers_number(layer_thick, layers_nb)

      !if (.NOT. ALLOCATED(layer_temp)) then
        allocate(layer_temp(layers_nb))
      !end if

      !if (.NOT. ALLOCATED(up_depth)) then
        allocate(up_depth(layers_nb + 1))
      !end if
     
      !if (.NOT. ALLOCATED(layer_depth)) then
        allocate(layer_depth(layers_nb))
      !end if

      up_depth = 0
     
     ! Getting layers bottom depth
     CALL layer_thickness2depth(layer_thick, layer_depth)

    
     DO z = 1, layers_nb
      
      depth_value = layer_depth(z)
      up_depth(z+1) = depth_value
     END DO

     ! Calculating layers mean temparature
     DO z = 1, layers_nb
      layer_temp(z) = sum(temp_profile((up_depth(z) + 1):up_depth(z + 1))) / &
                      layer_thick(z)
     END DO 
     !%%CyML Compute End%%
   END SUBROUTINE model_layers_temp
   !%%CyML Model End%%


   SUBROUTINE layer_thickness2depth(layer_thick, layer_depth)

    INTEGER, intent(in)     :: layer_thick(:)
    INTEGER, allocatable, intent(out) :: layer_depth(:)

    INTEGER :: layers_nb, z

    layers_nb = size(layer_thick) 

    !if (.NOT. ALLOCATED(layer_depth)) then
    allocate(layer_depth(layers_nb))
    !end if
    layer_depth = 0

    DO z = 1, layers_nb
       if (layer_thick(z) /= 0) layer_depth(z) = sum(layer_thick(1:z))
    END DO

   END SUBROUTINE layer_thickness2depth

   SUBROUTINE get_soil_depth(layer_thick, soil_depth)
    IMPLICIT NONE

    INTEGER, intent(in)     :: layer_thick(:)
    INTEGER, intent(out)    :: soil_depth
    
    soil_depth = sum(layer_thick)


   END SUBROUTINE get_soil_depth


   SUBROUTINE get_layers_number(layer_thick_or_depth, layers_number)
    IMPLICIT NONE

    INTEGER, intent(in)     :: layer_thick_or_depth(:)
    INTEGER, intent(out)    :: layers_number  
    
    integer :: z

    layers_number = 0
    DO z = 1, size(layer_thick_or_depth)
      IF(layer_thick_or_depth(z) /= 0) layers_number = layers_number + 1
    END DO
    
   END SUBROUTINE get_layers_number
   
END MODULE layers_temp_mod

File: Models/Stics_soil_temperature/src/cs/Stics_soil_temperature/Layers_temp.cs
using System;
using System.Collections.Generic;
using System.Linq;
public class layers_temp
{
    private int[] _layer_thick;
    public int[] layer_thick
        {
            get { return this._layer_thick; }
            set { this._layer_thick= value; } 
        }
        public layers_temp() { }
    
    public void  CalculateModel(soil_tempState s, soil_tempState s1, soil_tempRate r, soil_tempAuxiliary a, soil_tempExogenous ex)
    {
        //- Name: layers_temp -Version: 1.0, -Time step: 1
        //- Description:
    //            * Title: layers mean temperature model
    //            * Authors: None
    //            * Reference: doi:http://dx.doi.org/10.1016/j.agrformet.2014.05.002
    //            * Institution: INRAE
    //            * ExtendedDescription: None
    //            * ShortDescription: None
        //- inputs:
    //            * name: temp_profile
    //                          ** description : soil temperature profile
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLEARRAY
    //                          ** len : 
    //                          ** max : 50.0
    //                          ** min : -50.0
    //                          ** default : 0.0
    //                          ** unit : degC
    //            * name: layer_thick
    //                          ** description : layers thickness
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //                          ** datatype : INTARRAY
    //                          ** len : 
    //                          ** max : 
    //                          ** min : 
    //                          ** default : 
    //                          ** unit : cm
        //- outputs:
    //            * name: layer_temp
    //                          ** description : soil layers temperature
    //                          ** datatype : DOUBLEARRAY
    //                          ** variablecategory : state
    //                          ** len : 
    //                          ** max : 50.0
    //                          ** min : -50.0
    //                          ** unit : degC
        double[] temp_profile = s.temp_profile;
        double[] layer_temp ;
        int z;
        int layers_nb;
        List<int> up_depth = new List<int>();
        List<int> layer_depth = new List<int>();
        int depth_value;
        layers_nb = get_layers_number(layer_thick);
        layer_temp = new double[ layers_nb];
        up_depth = new List<int>(layers_nb + 1);
        layer_depth = new List<int>(layers_nb);
        for (var i = 0; i < layers_nb + 1; i++){up_depth.Add(0);}
        layer_depth = layer_thickness2depth(layer_thick);
        for (z=1 ; z!=layers_nb + 1 ; z+=1)
        {
            depth_value = layer_depth[z - 1];
            up_depth[z + 1 - 1] = depth_value;
        }
        for (z=1 ; z!=layers_nb + 1 ; z+=1)
        {
            layer_temp[z - 1] = temp_profile.ToList().GetRange((up_depth[z - 1] + 1 - 1),up_depth[(z + 1 - 1)]).Sum() / layer_thick[(z - 1)];
        }
        s.layer_temp= layer_temp;
    }
    public static List<int> layer_thickness2depth(int[] layer_thick)
    {
        List<int> layer_depth = new List<int>();
        int layers_nb;
        int z;
        layers_nb = layer_thick.Length;
        layer_depth = new List<int>(layers_nb);
        for (var i = 0; i < layers_nb; i++){layer_depth.Add(0);}
        for (z=1 ; z!=layers_nb + 1 ; z+=1)
        {
            if (layer_thick[z - 1] != 0)
            {
                layer_depth[z - 1] = layer_thick.ToList().GetRange(1 - 1,z).Sum();
            }
        }
        return layer_depth;
    }
    public static int get_layers_number(int[] layer_thick_or_depth)
    {
        int layers_number;
        int z;
        layers_number = 0;
        for (z=1 ; z!=layer_thick_or_depth.Length + 1 ; z+=1)
        {
            if (layer_thick_or_depth[z - 1] != 0)
            {
                layers_number = layers_number + 1;
            }
        }
        return layers_number;
    }
}